Git/commit.sublime-syntax

```diff
@@ -0,0 +1,84 @@
+%YAML 1.2
+---
+# Highlight regular git commits, merge commits, and tags.
+
+name: Git Commit
+file_extensions:
+  - COMMIT_EDITMSG
+  - MERGE_MSG
+  - TAG_EDITMSG
+  - syntax_test_git_commit
+  - syntax_test_git_merge
+  - syntax_test_git_tag
+scope: source.gitcommit
+
+contexts:
+  prototype:
+    - include: comments
+
+  main:
+    - include: commit-message
+
+  branch-line:
+    - match: \bOn branch\s+(.*)
+      captures:
+        1: support.type.branch-name.gitcommit
+
+  change:
+    - match: (new file|modified|deleted)\s*(:)\s*(.*)
+      captures:
+        1: keyword.change-list.change.gitcommit
+        2: punctuation.separator.change.gitcommit
+        3: keyword.change-list.file.gitcommit
+
+  change-list:
+    - include: change-list-header
+    - include: change
+
+  change-list-header:
+    - match: '\bChanges to be committed\s*(:)'
+      scope: keyword.change-list.heading.gitcommit
+      captures:
+        1: punctuation.separator.change-list.gitcommit
+
+  comments:
+    - match: '#'
+      scope: punctuation.definition.comment.gitcommit
+      push:
+        - meta_scope: comment.line.gitcommit
+        - match: \n
+          pop: true
+        - include: change-list
+        - include: branch-line
+        - include: date-line
+
+  commit-message:
+    # this absurd regex is necessary to:
+    #   1. match comments in the subject line
+    #   2. only style the first 50 chars in the subject line
+    - match: ([^#\n]{,50})[^#\n]*((#).*)?
+      captures:
+        1: string.unquoted.subject-line.gitcommit markup.heading.gitcommit
+        2: comment.line.gitcommit
+        3: punctuation.definition.comment.gitcommit
+      push:
+        # enforce empty line following subject line
+        - match: \n
+          push:
+            - match: \n
+              pop: true
+            - match: '.'
+              scope: invalid.missing-newline.gitcommit
+        # everything left that isn't a comment is the commit message
+        - match: '[^#]*'
+          scope: string.unquoted.commit-message.gitcommit
+          push:
```

Comment:
When do we pop from this push?

Reply:
Good question! This is kinda hacky. The goal is to enforce the empty-line-after-subject rule This is done with the previous `match: \n` context, which matches text between two newlines. This will actually highlight all even-numbered lines bright red. But we only want to highlight the _second_ line red (if there's any text there). The solution I came up with is to push a context on the stack that matches all remaining text in the commit message, before `match: \n` has a chance to match again. To ensure that `match: \n` can't match any more text, I push a context on the stack and never pop it. I actually use this trick in the parent context too, which is also never popped.

#### vim's highlighting of commits
![vim_commit](https://user-images.githubusercontent.com/1250684/30006724-cd7251aa-90b3-11e7-9711-18f08c62b78f.png)